mabwiser/treebandit.py

```diff
@@ -0,0 +1,175 @@
+
+    def predict_expectations(self, contexts: np.ndarray = None) -> Dict[Arm, Num]:
+
+        return self._parallel_predict(contexts, is_predict=False)
+
+    def _fit_arm(self, arm: Arm, decisions: np.ndarray, rewards: np.ndarray, contexts: Optional[np.ndarray] = None):
+
+        # Create dataset for the given arm
+        arm_contexts = contexts[decisions == arm]
+        arm_rewards = rewards[decisions == arm]
+
+        # Check that the dataset for the given arm is not empty
+        if arm_contexts.size != 0:
+
+            # Fit decision tree of the arm with this dataset
+            self.arm_to_tree[arm].fit(arm_contexts, arm_rewards)
+
+            # For each leaf, keep a list of rewards
+            # DecisionTreeClassifier's apply() method returns the indices of the nodes in the tree
+            # that the specified contexts lead to. Therefore, the indices returned are not necessarily
+            # consecutive/follow numerical order, but are always representative of leaf nodes.
+            leaf_indices = self.arm_to_tree[arm].apply(arm_contexts)
+
+            # Use set() to create a list of unique indices
+            # These indices represent the leaves reached via the given contexts
+            unique_leaf_indices = set(leaf_indices)
+
+            for index in unique_leaf_indices:
+                # Get rewards list for each leaf
+                self.arm_to_rewards[arm][index] = arm_rewards[leaf_indices == index]
+        else:
+            self.unfitted_arms.append(arm)
+
+    def _predict_contexts(self, contexts: np.ndarray, is_predict: bool,
+                          seeds: Optional[np.ndarray] = None, start_index: Optional[int] = None) -> List:
+
+        # Get local copy of arm_to_tree, arm_to_expectation, arm_to_rewards, and arms
+        # to minimize communication overhead between arms (processes) using shared objects
+        arm_to_tree = deepcopy(self.arm_to_tree)
+        arm_to_rewards = deepcopy(self.arm_to_rewards)
+        arm_to_expectation = deepcopy(self.arm_to_expectation)
+        arms = deepcopy(self.arms)
+
+        # Create an empty list of predictions
+        predictions = [None] * len(contexts)
+        for index, row in enumerate(contexts):
+            # Each row needs a separately seeded rng for reproducibility in parallel
+            rng = np.random.RandomState(seed=seeds[index])
+
+            for arm in arms:
+
+                # Copy the row rng to the deep copied model in arm_to_tree
+                arm_to_tree[arm].rng = rng
+
+                # If there was prior data for this arm, get expectation for arm
+                if arm_to_rewards[arm]:
+                    # Get leaf index for that context
+                    leaf_index = arm_to_tree[arm].apply([row])[0]
+
+                    # Get the rewards list of that leaf
+                    leaf_rewards = arm_to_rewards[arm][leaf_index]
+
+                    # Create leaf lp
+                    leaf_lp = self._get_leaf_lp(arm)
```

Comment:
So for every arm you are creating a new instance of the learning policy fit only to that arm?

Reply:
Yes. Specifically, we receive a context and follow it to a leaf in each arm's tree. Then, a new lp object is created for those leaves.

Reply:
So how does it create a softmax distribution or leverage epsilon for e-greedy?

Reply:
Oh, I see your point! I guess all TreeBandit does is use the learning policy algorithms to choose an expected reward for each arm. But given that we're only using the lps on one arm at a time, some lps may not be used to their full potential.